### PR TITLE
fix(engine): prevent L095 false positives on EDA rulebooks

### PR DESCRIPTION
**Pull request by ffirg**
## Summary
- EDA (Event-Driven Ansible) rulebooks were incorrectly detected as playbooks, causing L095 to flag valid EDA keywords (`sources`, `rules`) as \"unknown play keyword(s)\"
- This fix adds EDA detection to exclude rulebooks from playbook validation rules
- Content-based detection is now strict enough to keep invalid playbooks with stray `sources`/empty inline `rules` on the playbook path so L095 still reports them
- The regex fallback now also recognizes typical nested EDA rule indentation when YAML parsing fails, so malformed rulebooks still avoid the playbook path

## Changes
- Add `RULEBOOK`/`RULESET` to `NodeType` enum (prepares for future E-series rules per REQ-012)
- Add `could_be_eda_rulebook()` and `is_eda_rulebook_path()` detection in `awx_utils.py` (relative and absolute path segments, separator normalization)
- Update `could_be_playbook()` to exclude EDA files
- Tighten content-based EDA heuristics so stray play-level `sources` keys and inline empty `rules` outside EDA paths are not misclassified as rulebooks
- Broaden the regex-only fallback to accept the typical 6-space `condition`/`action` indentation used under EDA `rules:` entries
- Add `could_be_playbook_detail()` and `could_be_eda_rulebook_detail()` safeguards in `finder.py`; exclude EDA from `could_be_taskfile()`
- Add `RULEBOOK`/`RULESET` to scannable types in `graph_scanner.py`
- Add comprehensive regression tests for path-based EDA detection, playbook-vs-EDA edge cases, invalid playbooks that must still reach L095, and malformed rulebooks that rely on the regex fallback

Detection is based on:
- **Path patterns**: Files under `rulebooks/` or `extensions/eda/` (relative or absolute paths)
- **Content patterns**: EDA-style `rules` entries with rule structure, while stray playbook keys remain eligible for L095

## Test plan
- [x] `tox -e lint` passes
- [x] `tox -e unit` passes
- [x] EDA rulebooks in `rulebooks/` directories no longer trigger L095
- [x] Invalid playbooks with stray `sources`/empty inline `rules` still remain on the playbook path for L095
- [x] Malformed EDA rulebooks with valid rule structure still stay off the playbook path via regex fallback
- [x] EDA files labeled `rulebook` (not `taskfile`) and excluded from `search_playbooks` / scan targets

### CODE REVIEW

File: src/apme_engine/engine/awx_utils.py [hunk 1]
```diff
@@ -6,6 +6,45 @@
 
 valid_playbook_re = re.compile(r"^\s*?-?\s*?(?:hosts|include|import_playbook):\s*?.*?$")
 
+# EDA rulebook detection: look for 'sources:' or 'rules:' at ruleset level
+# These keys are valid in EDA rulebooks but not in Ansible playbooks
+_eda_rulebook_re = re.compile(r"^\s{2,4}(?:sources|rules):\s*$")
```
[the remark this answers is not in the document]
Comment (ffirg) [reply src/apme_engine/engine/awx_utils.py]: Fixed. Expanded `_eda_rulebook_re` to `r"^\s{1,8}(?:sources|rules):\s*(?:\S.*)?$"` which now allows 1-8 leading spaces (broader indentation) and optional trailing content (inline values like `[]`, comments, etc.). Added `test_eda_inline_content_detection` to verify this works. Fixed in a1ea2d3.

File: src/apme_engine/engine/awx_utils.py [hunk 2]
```diff
@@ -6,6 +6,45 @@
...
+    """
+    basename, ext = os.path.splitext(fpath)
+    if ext not in [".yml", ".yaml"]:
+        return False
+
+    # Path-based detection: files in rulebooks/ or extensions/eda/ directories
+    path_lower = fpath.lower()
+    if "/rulebooks/" in path_lower or "/extensions/eda/" in path_lower:
+        return True
+
```
[the remark this answers is not in the document]
Comment (ffirg) [reply src/apme_engine/engine/awx_utils.py:225]: Fixed. Updated the docstring to explicitly describe the two-tier detection approach: (1) path-based for /rulebooks/ and /extensions/eda/ directories, (2) content-based for files outside those directories. The docstring now clarifies that path-based detection is intentional and will classify any YAML under those directories as EDA. Fixed in a1ea2d3.

File: tests/test_eda_graph_integration.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+class TestEdaGraphIntegration:
+    """Tests that EDA rulebooks are handled correctly by the graph system."""
+
+    def test_eda_rulebook_excluded_from_playbook_detection(self, tmp_path: Path) -> None:
+        """Verify EDA rulebook files are not detected as playbooks.
+
+        EDA rulebooks have 'sources' and 'rules' keys which are not valid
+        playbook keywords. These files should be excluded from playbook
+        detection to prevent L095 false positives.
+
```
[the remark this answers is not in the document]
Comment (ffirg) [reply tests/test_eda_graph_integration.py:20]: Added `test_search_playbooks_excludes_eda_rulebooks` which exercises the actual entry point for project scanning. The test creates a mixed directory with playbooks and EDA rulebooks (both content-based and path-based detection), then verifies `search_playbooks()` returns only the playbook. Since `search_playbooks` calls `could_be_playbook` which calls `could_be_eda_rulebook`, EDA files are excluded before graph construction, preventing L095 from ever seeing them. Fixed in a1ea2d3.

File: src/apme_engine/engine/awx_utils.py [hunk 4]
```diff
@@ -6,6 +6,52 @@
...
+    Returns:
+        True if the file appears to be an EDA rulebook.
+    """
+    basename, ext = os.path.splitext(fpath)
+    if ext not in [".yml", ".yaml"]:
+        return False
+
+    # Path-based detection: files in rulebooks/ or extensions/eda/ directories
+    path_lower = fpath.lower()
+    if "/rulebooks/" in path_lower or "/extensions/eda/" in path_lower:
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Added `is_eda_rulebook_path()` with separator normalization and checks for both relative (`rulebooks/foo.yml`) and absolute (`/rulebooks/`) path segments, plus `extensions/eda/` variants. Covered by new unit tests. Fixed in cc9d6d9.

File: src/apme_engine/engine/finder.py [hunk 5]
```diff
@@ -743,6 +743,10 @@ def could_be_playbook_detail(body: str = "", data: YAMLValue | None = None, fpat
     if not isinstance(data[0], dict):
         return False
 
+    # EDA rulebooks have 'sources' or 'rules' keys - these are not playbooks
+    if "sources" in data[0] or "rules" in data[0]:
+        return False
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/finder.py]: Added `could_be_eda_rulebook_detail()`, excluded EDA from `could_be_taskfile()`, and `label_yml_file()` now returns `rulebook` (checked before playbook/taskfile). `list_scan_target()` already skips non-playbook/taskfile labels, so rulebooks stay out of scan targets. Fixed in cc9d6d9.

File: src/apme_engine/engine/finder.py [hunk 6]
```diff
@@ -743,12 +743,48 @@ def could_be_playbook_detail(body: str = "", data: YAMLValue | None = None, fpat
...
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/finder.py]: Removed the blanket `sources`/`rules` exclusion from `could_be_playbook_detail()`. Playbook loading in `model_loader` now gates on `could_be_eda_rulebook()` explicitly, so playbook-shaped YAML with unknown keys is not silently skipped. Fixed in 2168271.

File: src/apme_engine/engine/finder.py [hunk 7]
```diff
@@ -797,6 +836,9 @@ def label_empty_file_by_path(fpath: str) -> str:
         'taskfile', 'playbook', or empty string.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/finder.py:839]: Updated `label_empty_file_by_path()` docstring to document the `rulebook` return value alongside taskfile/playbook/empty string. Fixed in 2168271.

File: src/apme_engine/engine/awx_utils.py [hunk 8]
```diff
@@ -6,6 +6,85 @@
 
 valid_playbook_re = re.compile(r"^\s*?-?\s*?(?:hosts|include|import_playbook):\s*?.*?$")
 
+# EDA rulebook detection: look for 'sources:' or 'rules:' at ruleset level
+# These keys are valid in EDA rulebooks but not in Ansible playbooks
+# Allow any trailing content (inline values like [], comments, etc.)
+_eda_rulebook_re = re.compile(r"^\s{1,8}(?:sources|rules):\s*(?:\S.*)?$")
```
[the remark this answers is not in the document]
Comment (ffirg) [reply src/apme_engine/engine/awx_utils.py]: Fixed in 03faacc. Tightened the heuristic:

1. Now requires seeing `- name:` list item first (ruleset definition)
2. Then matches `sources:`/`rules:` at exactly 2-space indent only

This avoids K8s `spec.rules:` (deeper indent) and playbook `vars: {rules: ...}` (also deeper indent). The 2-space requirement matches ruleset-level keys that are direct children of the list item marker.

File: src/apme_engine/engine/model_loader.py [hunk 9]
```diff
@@ -1120,7 +1120,11 @@ def load_playbooks(
         if fpath in loaded:
             continue
 
-        if could_be_playbook(fpath=fpath) and could_be_playbook_detail(fpath=fpath):
+        if (
+            could_be_playbook(fpath=fpath)
+            and could_be_playbook_detail(fpath=fpath)
+            and not could_be_eda_rulebook(fpath=fpath)
+        ):
```
[the remark this answers is not in the document]
Comment (ffirg) [reply src/apme_engine/engine/model_loader.py]: Fixed in 03faacc. Removed the redundant check since `could_be_playbook()` already calls `could_be_eda_rulebook()` internally and returns False for EDA files. Added a clarifying comment explaining this.

File: src/apme_engine/engine/finder.py [hunk 10]
```diff
@@ -749,6 +749,38 @@ def could_be_playbook_detail(body: str = "", data: YAMLValue | None = None, fpat
...
+    if not body or not data:
+        return False
+
+    if not isinstance(data, list) or len(data) == 0:
+        return False
+
+    if not isinstance(data[0], dict):
+        return False
+
+    return "sources" in data[0] or "rules" in data[0]
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/finder.py]: Added `looks_like_eda_ruleset()` and wired `could_be_eda_rulebook_detail()` to use it: reject playbook section keys (`tasks`, `roles`, etc.), accept `sources` lists, and require `condition`/`action` on `rules` entries so invalid play-level `rules`/`sources` stay playbooks for L095. Fixed in f32250c.

File: src/apme_engine/engine/awx_utils.py [hunk 11]
```diff
@@ -6,6 +6,97 @@
...
+        saw_list_item = False
+        with codecs.open(fpath, "r", encoding="utf-8", errors="ignore") as f:
+            for n, line in enumerate(f):
+                if n > 100:
+                    break
+                if _eda_list_item_re.match(line):
+                    saw_list_item = True
+                elif saw_list_item and _eda_ruleset_key_re.match(line):
+                    return True
+    except OSError:
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Content-based `could_be_eda_rulebook()` now parses YAML when possible and applies the same `looks_like_eda_ruleset()` checks; regex fallback rejects playbook sections and requires `sources:` or `rules:` plus `condition`/`action` structure. Fixed in f32250c.

File: src/apme_engine/engine/awx_utils.py [hunk 12]
```diff
@@ -6,6 +6,97 @@
...
+    from being misclassified as playbooks (which would cause L095 false
+    positives for 'unknown play keyword' on sources/rules).
+
+    Args:
+        fpath: Path to the file to check.
+
+    Returns:
+        True if the file appears to be an EDA rulebook.
+    """
+    basename, ext = os.path.splitext(fpath)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Replaced unused `basename` bindings with `_` in `could_be_eda_rulebook()` and `could_be_playbook()`. Fixed in f32250c.

File: src/apme_engine/engine/awx_utils.py [hunk 13]
```diff
@@ -21,9 +189,14 @@ def could_be_playbook(fpath: str) -> bool:
...
     """
-    basename, ext = os.path.splitext(fpath)
+    _, ext = os.path.splitext(fpath)
     if ext not in [".yml", ".yaml"]:
         return False
+
+    # EDA rulebooks should not be treated as playbooks
+    if could_be_eda_rulebook(fpath):
+        return False
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py:253]: Restructured `could_be_playbook()` to read each file once: path-based EDA short-circuits without I/O, then the cheap hosts/import/vault regex runs first (vars/manifests return False immediately), and YAML parse runs only for playbook-shaped files that may contain ruleset-level `sources:`/`rules:`. Shared line helpers are used by `could_be_eda_rulebook()` too. Fixed in 5b5957f.

File: src/apme_engine/engine/awx_utils.py [hunk 14]
```diff
@@ -1,11 +1,233 @@
...
+    Args:
+        fpath: Path to the YAML file.
+        max_lines: Maximum number of lines to read.
+
+    Returns:
+        Lines read from the file, or None on I/O error.
+    """
+    try:
+        with open(fpath, encoding="utf-8", errors="ignore") as f:
+            return [line for n, line in enumerate(f) if n < max_lines]
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Replaced the list comprehension with a loop that breaks once `max_lines` is reached, so large YAML files are not read past the header window. Fixed in 0cb5abf.

File: src/apme_engine/engine/awx_utils.py [hunk 15]
```diff
@@ -1,11 +1,238 @@
...
+    if "sources" in ruleset:
+        sources = ruleset["sources"]
+        if sources is None or isinstance(sources, list):
+            return True
+    if "rules" in ruleset:
+        rules = ruleset["rules"]
+        if isinstance(rules, list):
+            if not rules:
+                return "sources" in ruleset
+            return any(isinstance(item, dict) and ("condition" in item or "action" in item) for item in rules)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Tightened `looks_like_eda_ruleset()` so content-based detection now requires EDA rule structure under `rules` instead of treating any `sources` list as an EDA match. I also added regressions for a playbook with stray `sources` before `tasks` and for content-only `sources` so those files stay on the playbook path and still reach L095. Fixed in f68b1e2.

File: src/apme_engine/engine/awx_utils.py [hunk 16]
```diff
@@ -1,11 +1,238 @@
...
+            return False
+        if _eda_list_item_re.match(line):
+            saw_list_item = True
+        elif saw_list_item and _eda_sources_key_re.match(line):
+            return True
+        elif saw_list_item and _eda_ruleset_key_re.match(line) and "rules:" in line:
+            saw_rules_key = True
+        elif saw_rules_key and _eda_rule_structure_re.match(line):
+            return True
+    return False
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Updated `_eda_from_lines()` so the `sources:` path no longer returns early; it now scans the rest of the header and lets later playbook-only sections like `tasks:` force a non-EDA result. The new stray-`sources` regression test locks in the case where `sources` appears before `tasks` but should still be validated as a playbook by L095. Fixed in f68b1e2.

File: src/apme_engine/engine/awx_utils.py [hunk 17]
```diff
@@ -1,11 +1,232 @@
...
+# EDA rulebook detection patterns (regex fallback when YAML does not parse):
+# 1. List item start: "- name:" at column 0 indicates ruleset definition
+# 2. Ruleset-level keys: "sources:" or "rules:" at exactly 2-space indent
+# 3. Playbook-only keys at 2-space indent reject EDA classification
+# 4. "rules:" alone requires a following "condition:" or "action:" at 4-space indent
+_eda_list_item_re = re.compile(r"^-\s+name:\s*\S")
+_eda_ruleset_key_re = re.compile(r"^  (?:sources|rules):\s*(?:\S.*)?$")
+_eda_sources_key_re = re.compile(r"^  sources:\s*(?:\S.*)?$")
+_eda_playbook_section_re = re.compile(r"^  (?:tasks|roles|handlers|pre_tasks|post_tasks):")
+_eda_rule_structure_re = re.compile(r"^    (?:condition|action):")
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/awx_utils.py]: Broadened `_eda_rule_structure_re` to accept the typical nested EDA rule-entry indentation range (`4-8` spaces, covering the usual 6-space `condition:`/`action:` layout) and added a regression for malformed YAML so the regex fallback still keeps rulebooks off the playbook path when parsing fails. Fixed in ab31f69.
